File: Leetcode_All_Solved/139.cpp
//top down DP
#define notyet (-1)

class Solution {
public:
    bool dfs(const string& s, int high, vector<int>& dp, const vector<string>& wordDict){
        if(high < 0)return true;
        if(dp[high] != notyet)return dp[high];

        bool ret = false;
        for(const string& word: wordDict){
            int low_tmp = high - (word.size()-1);
            if(low_tmp < 0)continue;
            if(word == s.substr(low_tmp, word.size()) )
                ret |= dfs(s, low_tmp-1, dp, wordDict);
        }

        return (dp[high] = ret);
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<int> dp(n, notyet);
        //NOTE: dp[i] := s[:i] satisfies the required segmentation
        return dfs(s, n-1, dp, wordDict);
    }
};

File: Leetcode_All_Solved/139.py
# O(n^2) for partition DP / DP Interval
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_set = set()
        for word in wordDict:
            word_set.add(word)
        
        @cache
        def enum_partition(i: int) -> bool:
            if(i==len(s)):
                return True
            
            subs = ""
            for j in range(i, len(s)):
                subs += s[j]
                if(subs not in word_set):
                    continue
                
                if(enum_partition(j+1) == True):
                    return True
            return False
        
        return enum_partition(0)
